90zfsbootmenu/zfsbootmenu-lib.sh

```diff
@@ -338,22 +338,32 @@ find_be_kernels() {
   : > "${kernel_records}"
 
   # shellcheck disable=SC2012,2086
-  for kernel in $( ls ${mnt}/boot/vmlinux-* \
-    ${mnt}/boot/vmlinuz-* \
-    ${mnt}/boot/kernel-* \
-    ${mnt}/boot/linux-* 2>/dev/null | sort -V ); do
-
-    kernel="${kernel#${mnt}}"
-    # shellcheck disable=SC2001
-    version=$( echo "$kernel" | sed -e "s,^[^0-9]*-,,g" )
-
-    for i in "initrd.img-${version}" "initrd-${version}.img" "initrd-${version}.gz" \
-      "initrd-${version}" "initramfs-${version}.img"; do
+  for kernel in $( ls \
+      ${mnt}/boot/{{vm,}linu{x,z},kernel}{,-*} 2>/dev/null | sort -V ); do
+    # Pull basename and validate
+    kernel=$( basename "${kernel}" )
+    [ -e "${mnt}/boot/${kernel}" ] || continue
+
+    # Kernel "base" extends to first hyphen
+    kernel_base="${kernel%%-*}"
+    # Kernel "version" is everything after base and may be empty
+    version="${kernel#${kernel_base}}"
+    version="${version#-}"
+
+    # initramfs images can take many forms, look for a sensible one
+    labels=( "$kernel" )
```

Comment:
What distribution does some flavor of `initramfs-linux.img` (or whatever this ends up matching in the big initramfs check defined later) ?

Reply:
It's been awhile since I ran Arch, but I'm pretty sure the default initramfs image for that distribution takes the forms
```
initramfs-linux.img
initramfs-linux-lts.img
initramfs-linux-zen.img
...
```
where the corresponding kernsl are `linux`, `linux-lts`, `linux-zen`, etc.

The [Arch wiki](https://wiki.archlinux.org/index.php/Mkinitcpio#Troubleshooting) seems to confirm this.

Reply:
Ahh, good to know. Would it make more sense then to reverse the ordering and set `labels=( "$version" )` first and then push `$kernel` onto the array? This way the Void naming pattern will be matched early in the cycle, rather than later.

Reply:
The advantage of keeping `$kernel` before `$version` is that it allows for kernel-specific images even when the kernels have the same version. In this totally contrived example, let's assume a user has installed `vmlinuz-5.7.7_1` and `linux-5.7.7_1` with different configurations. If we favor the default Void naming, then `initramfs-5.7.7_1.img` will be preferred for both kernels. However, if the user wants custom behavior for `linux-5.7.7_1`, preferring the `$kernel` suffix allows creation of `initramfs-linux-5.7.7_1.img` that will take priority with that kernel.

To improve matters quite a bit, I've unrolled most of the brace expansions into nested `for` loops that adjust the loop priorities. Now the Void default will hit third rather than the ~30th place (!!) it was hitting with the original expansion.